---
title: "Processing RNA-seq data"
author: "Ricardo Gomez-Reyes"
date: "9/3/2019"
output: word_document
---

```{r setup, include=FALSE}

# Setting global option in all the chunks-code
knitr::opts_chunk$set(warning=FALSE, echo = FALSE, 
                      message = FALSE, dpi=300,
                      fig.width=12, fig.height=8)

rm(list=ls())
```

```{r paths and functions, echo = FALSE}
dir <- "/Users/cigom/transcriptomics/oyster_full_assembly/"
name <- 'RSEM.isoforms.counts.matrix'

ctb.name <- "diffExp/RSEM.isoforms.counts.matrix"
metadta <- 'diffExp/metadata.file'

# annotation
annot.name <- 'annotation/Trinotate.xls'

cv_prot_name <- 'annotation/GCF_002022765.2_C_virginica-3.0_protein.faa.annot'
cv_nc_name <- 'annotation/GCF_002022765.2_C_virginica-3.0_rna.fna.annot'

cg_prot_name <- 'annotation/Crassostrea_gigas.oyster_v9.pep.all.fa.annot'
cg_nc_name <- 'annotation/Crassostrea_gigas.oyster_v9.cds.all.fa.annot'

setwd(dir)

# ==== Create a modified function
split_blast <- function (x, hit = "sprot_Top_BLASTX_hit")
{
    y <- x[!is.na(get(hit)), .(get(hit), gene_id, transcript_id,
        prot_id)]
    z <- strsplit(y$V1, "`")
    n <- sapply(z, length)
    z <- strsplit(unlist(z), "\\^")
    if (any(sapply(z, "[", 1) != sapply(z, "[", 2)))
        print("WARNING: check different values in columns 1 and 2")
    NAME <- gsub("^RecName: Full=", "", sapply(z, "[", 6))
    NAME <- gsub("SubName: Full=", "", NAME)
    NAME <- gsub(";$", "", NAME)
    NAME <- gsub(" \\{[^}]+}", "", NAME)
    x1 <- data.frame(gene = rep(y$gene_id, n), transcript = rep(y$transcript_id,
        n), protein = rep(gsub(".*\\|", "", y$prot_id), n), uniprot = sapply(z,
        "[", 1), align = sapply(z, "[", 3), identity = as.numeric(gsub("%ID",
        "", sapply(z, "[", 4))), evalue = as.numeric(gsub("E:",
        "", sapply(z, "[", 5))), name = NAME, lineage = sapply(z,
        "[", 7), domain = gsub("; .*", "", sapply(z, "[", 7)),
        genus = gsub(".*; ", "", sapply(z, "[", 7)), stringsAsFactors = FALSE)
    message(nrow(x1), " ", hit, " annotations")
    data.table(x1)
}

# gene ontology pfam
split_GOp <- function (x, hit = "gene_ontology_pfam") 
{
    y <- x[!is.na(get(hit)), .(get(hit), gene_id, transcript_id, 
        prot_id)]
    z <- strsplit(y$V1, "`")
    n <- sapply(z, length)
    z <- strsplit(unlist(z), "\\^")
    x1 <- data.frame(gene = rep(y$gene_id, n), transcript = rep(y$transcript_id, 
        n), protein = rep(gsub(".*\\|", "", y$prot_id), n), go = sapply(z, 
        "[", 1), ontology = sapply(z, "[", 2), name = sapply(z, 
        "[", 3), stringsAsFactors = FALSE)
    message(nrow(x1), " ", hit, " annotations")
    data.table(x1)
}

```

## Cargamos los datos:

```{r read count-table, echo = FALSE}

# file <- list.files(path = ctb.name, full.names = TRUE)
file <- paste0(dir, ctb.name)

dim(count_tbl <- read.csv(file, header=T, com='', row.names=1, check.names=F, sep='\t', stringsAsFactors = FALSE))

```

```{r read groups- metadata, echo = FALSE}
metadata.file <- paste0(dir, metadta)

metadata <- read.csv(metadata.file, header=T, 
                     com='', check.names=F, 
                     sep='\t', stringsAsFactors = FALSE)


identical(names(count_tbl), metadata$factor2)
rep <- metadata$Replica
```

## Filtrado de matriz de conteos
```{r 1. mean replicates, echo = FALSE}
dim(data <- count_tbl) # restore before doing various data transformations
# group by replicate-mean
#names(data) <- rep
names(data) <- metadata$Replica

#select(metadata, Identificador, factor1)

# using rowMeans
dim(data <- sapply(unique(rep), function(g) rowMeans(data[,rep==g,drop=FALSE])))

```

```{r 2. remove inactive - genes}
dim(data)
#dim(data_sbt <- data[rowMeans(data) >= 5,])
#dim(data_sbt <- rpm[rownames(rpm) %in% rownames(data[rowMeans(data) >= 5,]),])
dim(data_sbt <- data[rownames(data) %in% rownames(data[rowMeans(data) >= 5,]),])

# log2(FPKM) values less than âˆ’15 were set to not-detected

```

## Parse annotation

```{r Load annotation}
library(trinotateR)
annot.file <- paste0(dir, annot.name)
x <- read_trinotate(annot.file)
summary_trinotate(x)
```

```{r Detect Ribosomal}
na.exclude(top_table(x, "RNAMMER", n = 10))

rnamer_id <- x[!is.na(x$RNAMMER),]$transcript_id
rnamer_id <- c(rnamer_id, 'TRINITY_DN53822_c1_g1_i1')
rname_abundance <- data_sbt[rownames(data_sbt) %in% rnamer_id,]
knitr::kable(t(rname_abundance *100))
# Remove ribosomal from annotation:
dim(x1 <- x[is.na(x$RNAMMER),])
# and from quantification
dim(data_sbt)
dim(data_sbt_1 <- data_sbt[!rownames(data_sbt) %in% rnamer_id,])

```

```{r GETBACK the genome feature}
cv_nc_file <- paste0(dir, cv_nc_name)
cv_prot_file <- paste0(dir, cv_prot_name)

cv_nc_annot <- read.csv(cv_nc_file, header = FALSE, sep = ' ')
cv_nc_annot <- data.frame(id = cv_nc_annot$V1, annot = apply(cv_nc_annot[-1], 1, paste, collapse=" "))

cv_prot_annot <- read.csv(cv_prot_file, header = FALSE, sep = ' ')
cv_prot_annot <- data.frame(id = cv_prot_annot$V1, annot = apply(cv_prot_annot[-1], 1, paste, collapse=" "))

# and for c gigas

cg_prot_annot <- read.csv(paste0(dir, cg_prot_name), header = FALSE, sep = ' ')
cg_nuc_annot <- read.csv(paste0(dir, cg_nc_name), header = FALSE, sep = ' ')

cg_prot_annot <- data.frame(id = cg_prot_annot$V1, annot = apply(cg_prot_annot[-c(1,2)], 1, paste, collapse=" "))

cg_nuc_annot <- data.frame(id = cg_nuc_annot$V1, annot = apply(cg_nuc_annot[-c(1,2)], 1, paste, collapse=" "))

```

```{r parse data}
library(dplyr)
pfam <- as.data.frame(split_pfam(x1))

go <- split_GO(x1)
go_pfam <- split_GO(x1, 'gene_ontology_pfam')

blastx <- split_blast(x1, "sprot_Top_BLASTX_hit")
blastp <- split_blast(x1, "sprot_Top_BLASTP_hit")

blastp_genome0 <- split_blast(x1, "GCFvirginica_blastp_BLASTP")
blastn_genome0 <- split_blast(x1, "GCFvirginica_blastn_BLASTX")

blastp_cgigas <- split_blast(x1, "CGigas_pep_BLASTP")
blastn_cgigas <- split_blast(x1, "CGigas_nuc_BLASTX")

names(blastp_genome0)[4] <- 'genome_id'
names(blastn_genome0)[4] <- 'genome_id'

cv_prot <-  cv_prot_annot[cv_prot_annot$id %in% blastp_genome0$genome_id,]
cv_prot <- cv_prot[match(blastp_genome0$genome_id, cv_prot$id),]

blastp_genome <- cbind(select(blastp_genome0, -name, -lineage, -domain, -genus), select(cv_prot, -id))
                       
cv_nc <-  cv_nc_annot[cv_nc_annot$id %in% blastn_genome0$genome_id,]
cv_nc <- cv_nc[match(blastn_genome0$genome_id, cv_nc$id),]

blastn_genome <- cbind(select(blastn_genome0, -name, -lineage, -domain, -genus), select(cv_nc, -id))
    
```

```{r 3. remove ribosomal from count and rename average samples}
data_sbt_1 <- as.data.frame(data_sbt[!rownames(data_sbt) %in% rnamer_id,])

rep_samples <- names(which(table(rep) > 1))

rename_p <- paste0(rep_samples, "_P")
  
rep_pos <- which(names(as.data.frame(data)) %in% rep_samples)

names(data_sbt_1)[rep_pos] <- rename_p
```

```{r define annotation - transcript level}
annot_matrix <- matrix(ncol = 3, nrow = length(rownames(data_sbt_1)))
annot_matrix[,1] <- rownames(data_sbt_1) %in%  unique(blastn_genome$transcript)
annot_matrix[,2] <- rownames(data_sbt_1) %in% unique(go$transcript)
annot_matrix[,3] <- rownames(data_sbt_1) %in%  unique(blastx$transcript)

annot_tbl <- data.frame(annot_matrix)
summary(annot_tbl)
# complate cases of annotation:
boolean.integer <- function(arg1) {
  arg1 [arg1 == "TRUE"] <- 1
  arg1 [arg1 == "FALSE"] <- 0
  arg1 <- as.integer(arg1)
}

annot_tbl_boo <- data.frame(apply(annot_tbl, 2, boolean.integer))
rownames(annot_tbl_boo) <- rownames(data_sbt_1)
names(annot_tbl_boo) <- c('cv_nuc','go', 'swissprot')

annot_tbl_boo$complete <- rowSums(annot_tbl_boo)
dim(complete_annot <- annot_tbl_boo[annot_tbl_boo$complete == 3,])

non_annotated <- rownames(annot_tbl_boo[annot_tbl_boo$complete == 0,])
dim(data_sbt_non_annotated <- data_sbt_1[rownames(data_sbt_1) %in% non_annotated, ])
# and filterd based on the mean fpkm >= 5
dim(data_sbt_non_annotated <- data_sbt_non_annotated[rownames(data_sbt_non_annotated) %in% rownames(data_sbt_non_annotated[rowMeans(data_sbt_non_annotated) >= 5,]),])

# And save complete cases:
dim(data_sbt_complete_cases_nuc <- data_sbt_1[rownames(data_sbt_1) %in% rownames(complete_annot), ])

# transcript-level
dim(z <- data_sbt_non_annotated)

dim(complete_cases_cv_nuc <- save_QA(data_sbt_complete_cases_nuc, blastn_genome, z))
dim(complete_cases_go_nuc <- save_QA(data_sbt_complete_cases_nuc, go,z))
dim(complete_cases_swissprot_nuc <- save_QA(data_sbt_complete_cases_nuc, blastx,z))

reorder_ <- function(x,y) {
  x <- x[match(rownames(y), x$transcript),]
  return(x)
  
  identical(rownames(y), x$transcript)
}

complete_cases_cv_nuc <- reorder_(complete_cases_cv_nuc, data_sbt_complete_cases_nuc)
complete_cases_go_nuc <- reorder_(complete_cases_go_nuc, data_sbt_complete_cases_nuc)
complete_cases_swissprot_nuc <- reorder_(complete_cases_swissprot_nuc, data_sbt_complete_cases_nuc)

# # sanity-check the ids per database
identical(rownames(data_sbt_complete_cases_nuc), complete_cases_cv_nuc$transcript)
identical(rownames(data_sbt_complete_cases_nuc), complete_cases_go_nuc$transcript)
identical(rownames(data_sbt_complete_cases_nuc), complete_cases_swissprot_nuc$transcript)

samples <- names(data_sbt_complete_cases_nuc)


dim(save_complete_nuc <- cbind(complete_cases_cv_nuc, 
                               select(complete_cases_go_nuc, 
                                      -samples, -gene, -transcript, -protein),
                               select(complete_cases_swissprot_nuc, 
                                      -samples, -gene, -transcript, -protein)))

write.table(save_complete_nuc, file = paste0(dir, name, '_cv_go_swissprot_nuc.txt'), sep="\t", 
            row.names = F, col.names = T)

```

```{r Define annotation - peptide level}
# =========
# orf level
# =========

annot_matrix <- matrix(ncol = 4, nrow = length(rownames(data_sbt_1)))
annot_matrix[,1] <- rownames(data_sbt_1) %in%  blastp_genome$transcript
annot_matrix[,2] <- rownames(data_sbt_1) %in% pfam$transcript
annot_matrix[,3] <- rownames(data_sbt_1) %in% go_pfam$transcript
annot_matrix[,4] <- rownames(data_sbt_1) %in% blastp$transcript

annot_tbl <- data.frame(annot_matrix)
summary(annot_tbl)
annot_tbl_boo_pep <- data.frame(apply(annot_tbl, 2, boolean.integer))

rownames(annot_tbl_boo_pep) <- rownames(data_sbt_1)
names(annot_tbl_boo_pep) <- c('cv_prot','pfam','go_pfam', 'swissprot')
# complate cases of annotation:

annot_tbl_boo_pep$complete <- rowSums(annot_tbl_boo_pep)

dim(complete_annot <- annot_tbl_boo_pep[annot_tbl_boo_pep$complete == 4,])

# get-back results
dim(data_sbt_complete_cases <- data_sbt_1[rownames(data_sbt_1) %in% rownames(complete_annot), ])

non_annotated <- rownames(annot_tbl_boo_pep[annot_tbl_boo_pep$complete == 0,])
dim(data_sbt_non_annotated <- data_sbt_1[rownames(data_sbt_1) %in% non_annotated, ])
# and filterd based on the mean fpkm >= 5
dim(data_sbt_non_annotated <- data_sbt_non_annotated[rownames(data_sbt_non_annotated) %in% rownames(data_sbt_non_annotated[rowMeans(data_sbt_non_annotated) >= 5,]),])

# And save complete cases:
dim(data_sbt_complete_cases_nuc <- data_sbt_1[rownames(data_sbt_1) %in% rownames(complete_annot), ])

# transcript-level
dim(z <- data_sbt_non_annotated)

dim(complete_cases_cv_pep <- save_QA(data_sbt_complete_cases, blastp_genome, z))
dim(complete_cases_pfam_pep <- save_QA(data_sbt_complete_cases, pfam, z))
dim(complete_cases_go_pep <- save_QA(data_sbt_complete_cases, go_pfam,z))
dim(complete_cases_swissprot_pep <- save_QA(data_sbt_complete_cases, blastp,z))


complete_cases_cv_pep <- reorder_(complete_cases_cv_pep, data_sbt_complete_cases)
complete_cases_pfam_pep <- reorder_(complete_cases_pfam_pep, data_sbt_complete_cases)
complete_cases_go_pep <- reorder_(complete_cases_go_pep, data_sbt_complete_cases)
complete_cases_swissprot_pep <- reorder_(complete_cases_swissprot_pep, data_sbt_complete_cases)


# # sanity-check the ids per database
identical(rownames(data_sbt_complete_cases), complete_cases_cv_pep$transcript)
identical(rownames(data_sbt_complete_cases), complete_cases_pfam_pep$transcript)
identical(rownames(data_sbt_complete_cases), complete_cases_go_pep$transcript)
identical(rownames(data_sbt_complete_cases), complete_cases_swissprot_pep$transcript)

samples <- names(data_sbt_complete_cases)


dim(save_complete_pep <- cbind(complete_cases_cv_pep, 
                               select(complete_cases_pfam_pep, 
                                      -samples, -gene, -transcript, -protein),
                               select(complete_cases_go_pep, 
                                      -samples, -gene, -transcript, -protein),
                               select(complete_cases_swissprot_pep,
                                      -samples, -gene, -transcript, -protein)))

write.table(save_complete_pep, file = paste0(dir, name, '_cv_pfam_goPfam_swissprot_pep.txt'), sep="\t", 
            row.names = F, col.names = T)
```

```{r parse annot & quantification}
# for nucleotide level:

dim(data_sbt_go <- data_sbt_1[rownames(data_sbt_1) %in% go$transcript, ])
dim(data_sbt_blastx <- data_sbt_1[rownames(data_sbt_1) %in% blastx$transcript, ])
dim(data_sbt_cv_nuc <- data_sbt_1[rownames(data_sbt_1) %in% blastn_genome$transcript, ])
# for orf level
dim(data_sbt_pfam <- data_sbt_1[rownames(data_sbt_1) %in% pfam$transcript, ])
dim(data_sbt_blastp <- data_sbt_1[rownames(data_sbt_1) %in% blastp$transcript, ])
dim(data_sbt_goPfam <- data_sbt_1[rownames(data_sbt_1) %in% go_pfam$transcript, ])
dim(data_sbt_cv_prot <- data_sbt_1[rownames(data_sbt_1) %in% blastp_genome$transcript, ])

# for C.Gigas genome
dim(data_sbt_cg_prot <- data_sbt_1[rownames(data_sbt_1) %in% blastp_cgigas$transcript, ])
dim(data_sbt_cg_nuc <- data_sbt_1[rownames(data_sbt_1) %in% blastn_cgigas$transcript, ])

```

```{r non-annotated but abundace}

annot_tbl <- cbind(select(annot_tbl_boo_pep, -complete), 
                   select(annot_tbl_boo, -complete))

annot_tbl$complete <- rowSums(annot_tbl)

non_annotated <- rownames(annot_tbl[annot_tbl$complete == 0,])

dim(data_sbt_non_annotated <- data_sbt_1[rownames(data_sbt_1) %in% non_annotated, ])

# check top-abundant of non-annotated data
# data_sbt_non_annotated %>% 
#   data.frame() %>%
#   top_n(25) -> non_annotated_included  
```

```{r test get unique function}
# keep the last annotation per dbs
# test
head(pfam$transcript[which(table(pfam$transcript) >= 3)])

tt <- pfam[pfam$transcript == 'TRINITY_DN18353_c0_g1_i1', c('transcript','pfam', 'symbol','name', 'align', 'evalue')]
tt[!duplicated(tt$transcript, fromLast = TRUE) ,]
# test finish

# keep the last transcript with highes evalue
# Save Quantitative Annotation
save_QA <- function(x,y, non_annot) {
  db_unique <- y[!duplicated(y$transcript, fromLast = TRUE) ,]
  a <- db_unique[db_unique$transcript %in% rownames(x),]
  a <- a[match(rownames(x), a$transcript),]
  
  if(identical(rownames(x), as.character(a$transcript))) {
    results <- cbind(x, a)
  }
  if(exists('non_annotated')) {
    z <- non_annot
    add <- matrix(ncol = ncol(z) + ncol(y), nrow = nrow(z))
    add[,1:ncol(z)] <- as.matrix(z)
    add <- as.data.frame(add)
    names(add) <- c(names(as.data.frame(z)), names(y))
    rownames(add) <- rownames(z)
    results0 <- rbind(results, add)
    
    return(results0)
    
    } else {
      results0 <- results
      return(results0)
    
  }
}


# x <- data_sbt_pfam
# z <- data_sbt_non_annotated
# y <- pfam
# add <- matrix(ncol = ncol(z) + ncol(y), nrow = nrow(z))
# add[,1:ncol(z)] <- as.matrix(z)
# add <- as.data.frame(add)
# names(add) <- c(names(as.data.frame(z)), names(y))
# rownames(add) <- rownames(z)
# 
# db_unique <- y[!duplicated(y$transcript, fromLast = TRUE) ,]
# a <- db_unique[db_unique$transcript %in% rownames(x),]
# 
# identical(rownames(x), as.character(a$transcript))
# a <- a[match(rownames(x), a$transcript),]
# identical(rownames(x), as.character(a$transcript))
# 
# results <- cbind(x, a)

```

```{r keep the last annotation per dbs}

#  ORF level
dim(save_pfam <- save_QA(data_sbt_pfam, pfam, data_sbt_non_annotated))
dim(save_blastp <- save_QA(data_sbt_blastp, blastp, data_sbt_non_annotated))
dim(save_goPfam <- save_QA(data_sbt_goPfam, go_pfam, data_sbt_non_annotated))
dim(save_cv_prot <- save_QA(data_sbt_cv_prot, blastp_genome, data_sbt_non_annotated))

# Nucleotide level
dim(save_blastx <- save_QA(data_sbt_blastx, blastx, data_sbt_non_annotated))
dim(save_go <- save_QA(data_sbt_go, go, data_sbt_non_annotated))
dim(save_cv_nuc <- save_QA(data_sbt_cv_nuc, blastn_genome, data_sbt_non_annotated))

# C gigas data
dim(save_cg_nuc <- save_QA(data_sbt_cg_nuc, blastn_cgigas, data_sbt_non_annotated))
dim(save_cg_prot <- save_QA(data_sbt_cg_prot, blastp_cgigas, data_sbt_non_annotated))


```

```{r save results}
write.table(save_pfam, file = paste0(dir, name, '.pfam_orfs.txt'), sep="\t", 
            row.names = F, col.names = T)

write.table(save_blastp, file = paste0(dir, name, '.swissProt_orfs.txt'), sep="\t", 
            row.names = F, col.names = T)

write.table(save_goPfam, file = paste0(dir, name, '.gene_ontology_orfs.txt'), sep="\t", 
            row.names = F, col.names = T)

#

write.table(save_blastx, file = paste0(dir, name, '.swissProt_nuc.txt'), sep="\t", 
            row.names = F, col.names = T)
#
write.table(save_go, file = paste0(dir, name, '.gene_ontology_nuc.txt'), sep="\t", 
            row.names = F, col.names = T)


# Genomes
write.table(save_cv_prot, file = paste0(dir, name, '.Cvirginica_orfs.txt'), sep="\t", 
            row.names = F, col.names = T)

write.table(save_cv_nuc, file = paste0(dir, name, '.Cvirginica_nuc.txt'), sep="\t", 
            row.names = F, col.names = T)

write.table(save_cg_nuc, file = paste0(dir, name, '.CGigas_nuc.txt'), sep="\t", 
            row.names = F, col.names = T)

write.table(save_cg_prot, file = paste0(dir, name, '.CGigas_orfs.txt'), sep="\t", 
            row.names = F, col.names = T)

```

## quit()

```{r}
# incluir ademas, version de genoma x/p de ostion en el trinotate.xls

# protein
#/LUSTRE/apps/bioinformatica/Trinotate/util/trinotateSeqLoader/Trinotate_BLAST_loader.pl --sqlite Trinotate.sqlite --outfmt6 GENOME_REFERENCE/good.Trinity.fasta.transdecoder_vs_GCF_002022765.2_C_virginica-3.0_protein_blastp.outfmt6 --prog blastp --dbtype GCFvirginica_blastp
# nuc
#/LUSTRE/apps/bioinformatica/Trinotate/util/trinotateSeqLoader/Trinotate_BLAST_loader.pl --sqlite Trinotate.sqlite --outfmt6 nuc_GENOME_REFERENCE/good.Trinity_vs_GCF_002022765.2_C_virginica-3.0_rna_blastn.outfmt6 --prog blastx --dbtype GCFvirginica_blastn
# Nuc c.gigas
#/LUSTRE/apps/bioinformatica/Trinotate/util/trinotateSeqLoader/Trinotate_BLAST_loader.pl --sqlite Trinotate.sqlite --outfmt6 nuc_GENOME_CGigas/good.Trinity_vs_Crassostrea_gigas.oyster_v9.cds.all_blastn.outfmt6 --prog blastx --dbtype CGigas_nuc
# pep c.gigas
#/LUSTRE/apps/bioinformatica/Trinotate/util/trinotateSeqLoader/Trinotate_BLAST_loader.pl --sqlite Trinotate.sqlite --outfmt6 pep_GENOME_CGigas/good.Trinity.fasta.transdecoder_vs_Crassostrea_gigas.oyster_v9.pep.all_blastp.outfmt6 --prog blastp --dbtype CGigas_pep

# re-escribimos el trinotate.xls
# Trinotate Trinotate.sqlite report > Trinotate.xls
```

### Some Visualization

```{r histogram of identity}


xy <- data.frame(Identity=blastp_genome$identity, Type="Transcript")
y <- data.frame(Identity=blastn_genome$identity, Type="ORF")

datavis <- rbind(xy,y)

save_plot <- ggplot(datavis, aes(Identity, fill = Type)) + 
        geom_histogram(bins = 100, alpha = 0.7, aes(y = ..density..), position = 'identity') + 
        theme_classic() +
        scale_fill_brewer(direction = -1, palette = "Paired") + 
        stat_function(fun=dnorm,
                     color="red",
                     args=list(mean=mean(datavis$Identity, na.rm = TRUE), 
                              sd=sd(datavis$Identity, na.rm = TRUE))) +  
        scale_x_continuous("Identity of the aligment vs genome")

png(paste0(dir,"/", "identity_trancript_vs_genome.png"), units="px", width=3700, height=2500, res=400)
plot(save_plot, col=adjustcolor("black", alpha=0.2))
dev.off()
```

```{r PcoA plot, echo = FALSE}
pca <- prcomp(t(count_tbl), center = TRUE, scale. = FALSE)

## make a scree plot
pca.var <- pca$sdev^2
pca.var.per <- round(pca.var/sum(pca.var)*100, 1)
#
pca.data <- data.frame(Sample=rownames(pca$x),
                       X=pca$x[,1],
                       Y=pca$x[,2])

pca.data <- cbind(pca.data, metadata)


n <-length(unique(pca.data$Sexo))
getPalette = colorRampPalette(brewer.pal(n, "Dark2"))


library(ggplot2)
pcaplot <- ggplot(data=pca.data, aes(x=X, y=Y, label=Grupo, color = Sexo)) +
  geom_text(label.size = 0.01) +
  geom_point() +
  #geom_point(aes(color = factor), size = 5, alpha = 0.7) +
  xlab(paste("PC1 - ", pca.var.per[1], "%", sep="")) +
  ylab(paste("PC2 - ", pca.var.per[2], "%", sep="")) + theme_bw() +
  scale_color_manual(values=getPalette(n)) + facet_wrap(~Tejido)
  
```

```{r mean visualization, echo = FALSE}
databox <- y 
names(databox) <- metadata$factor1
datam <- reshape2::melt(databox, value.name = 'rawCount')

# sapply(strsplit(as.character(datam$variable), "_"), `[`, 2)
  
p1 <- ggplot(datam, aes(x=variable, y=rawCount)) + 
  geom_boxplot() + facet_wrap(~condition, scales = 'free', nrow = 1) +
  stat_summary(fun.y=mean, geom="line", color = 'red', mapping = aes(group = condition)) +
  labs(title="No transformed data count",
       x="Samples (replicates)", 
       y = "Expression") +
  theme_classic() 


# test means after data-transformation ----

```

```{r scale to reads per million (rpm)}
# barplot(colSums(data))
rpm <- apply(data, 2, function(x) {x / sum(x)})
# sanity check
barplot(colSums(rpm))
```

```{r -log2 transformation, echo = FALSE}
#data <- y[-1] # restore before doing various data transformations
dataExp <- log2(data+1)
dataExp <- as.matrix(dataExp) # convert to matrix
dataExp <- t(scale(t(dataExp), scale=F)) # Centering rows
dataExp <- as.data.frame(dataExp)

datam2 <- reshape2::melt(data, value.name = 'log2')
datam2$rep <- sapply(strsplit(as.character(datam2$variable), "_"), `[`, 2)
datam2$condition <- sapply(strsplit(as.character(datam2$variable), "_"), `[`, 1)

library(ggplot2)

p2 <- ggplot(datam2, aes(x=rep, y=log2)) + 
  geom_boxplot() + facet_wrap(~condition, scales = 'free', nrow = 1) +
  stat_summary(fun.y=mean, geom="line", color = 'red', mapping = aes(group = condition)) +
  labs(title="log2 Transformed data count",x="Samples (replicates)", y = "Expression") +
  theme_classic() 

png(paste0(dir, "boxplot.png"), units="px", width=2900, height=2800, res=400)
gridExtra::grid.arrange(p1,p2)
dev.off()
```

```{r -datavis-time demand}
.cran_packages <- c('dplyr', 'superheat', 'ggplot2', 'RColorBrewer')
.inst <- .cran_packages %in% installed.packages()

if(any(!.inst)) {
  install.packages(.cran_packages[!.inst], dep=TRUE, repos='http://cran.us.r-project.org')
}

sapply(c(.cran_packages), require, character.only = TRUE)

# visualizing ----

# png(paste0(path_s, "Heatmap3_1.png"), units="px", width=2200, height=3000, res=400)

superheat(dataExp,
          #yr = yr,
          #yr.axis.name = "log2(Expr) \n",
          #yr.plot.t = "scattersmooth", 
          #yr.axis.name.angle = 90,
          #yr.axis = FALSE,
          
          # include barplot of R.A of those genes
          #yt = yt,
          #yt.plot.type = "box",
          #yt.axis.name = "Log2 (Expr)",
          #yr.cluster.col = 
          
          # retain original order of rows/cols
          pretty.order.rows = TRUE,
          pretty.order.cols = TRUE,
          #row.dendrogram = TRUE,
          # left labels
          left.label.size = 0.5,
          left.label.text.size = 1.7,
          left.label.text.alignment = 'right',
          #left.label.text.angle = 95,
          # bottom labels
          bottom.label.size = 0.5,
          bottom.label.text.size = 2.5,
          bottom.label.text.angle = 75,
          bottom.label.text.alignment = "right",
          row.title = " ",
          column.title = " ",
          # change the grid color
          grid.hline.col = "white",
          #grid.vline.col = "white",
          left.label.col = "white",
          bottom.label.col = "white",
          heat.pal = brewer.pal(n = 9, name = "RdYlBu"),
          force.left.label = TRUE,
          legend = TRUE,
          legend.vspace = -0.12,
          # make the legend bigger
          legend.height = 0.2,
          legend.width = 1,
          legend.text.size = 7
)
```

```{r Gaussian fit and zFPKM normalization}
#  https://bioconductor.org/packages/devel/bioc/vignettes/zFPKM/inst/doc/zFPKM.html
# https://bmcgenomics.biomedcentral.com/articles/10.1186/1471-2164-14-778#MOESM2
# It works well for gene level data using FPKM or TPM, but does not appear to calibrate well for transcript level data.
```

```{r analisis of cov (ANOVA)}
plot(log10(apply(data_sbt, 1, function(x){var(x)})), log10(rowMeans(data_sbt)),
     xlab = 'variation',
     ylab = 'mean')
plot(apply(data_sbt, 1, function(x){var(x)}), 
     rowMeans(data_sbt),
     xlab = 'variation',
     ylab = 'mean')

# per sample box

boxplot(apply(data_sbt, 2, function(x){var(x)}))

```

```{r check n of annotations}
# binary heatmap?
boolean.integer <- function(arg1) {
  arg1 [arg1 == "TRUE"] <- 1
  arg1 [arg1 == "FALSE"] <- 0
  arg1 <- as.integer(arg1)
}

annot_tbl_boo <- apply(annot_tbl, 2, boolean.integer)
superheat(annot_tbl_boo)



```



```bash
# es necesario instalar hmmsearch2 version antigua para usar el flag compact!
tar zxvf hmmer-2.3.tar.gz

cd hmmer-2.3

./configure --mandir=`pwd` --bindir=`pwd`/bin --prefix=`pwd`

make
make check
make install

# or download compiled version

wget http://eddylab.org/software/hmmer/2.3/hmmer-2.3.bin.intel-linux.tar.gz .

# then edit
grep '$HMMSEARCH_BINARY' rnammer

export PATH=/LUSTRE/bioinformatica_data/genomica_funcional/bin/RNAMMER2_TRINOTATE/hmmer-2.3.bin.intel-linux/binaries/:$PATH


rnammer=/LUSTRE/bioinformatica_data/genomica_funcional/bin/rnammer/
$rnammer/rnammer -S euk -m tsu,lsu,ssu -gff tmp.superscaff.rnammer.gff < transcriptSuperScaffold.fasta

which: no hmmsearch2 in ($PATH)

Error, cannot find path to hmmsearch at /LUSTRE/bioinformatica_data/genomica_funcional/bin/rnammer/rnammer line 53.

Error, cmd: perl 

$rnammer/rnammer -S euk -m tsu,lsu,ssu -gff tmp.superscaff.rnammer.gff < transcriptSuperScaffold.fasta died with ret 65280 at 


# =====
HMMSEARCH_BINARY=/LUSTRE/bioinformatica_data/genomica_funcional/bin/RNAMMER2_TRINOTATE/hmmer-2.3.bin.intel-linux/binaries/
export PATH=$HMMSEARCH_BINARY:$PATH


SOURCE=/LUSTRE/bioinformatica_data/genomica_funcional/scripts/exports

source $SOURCE

export PATH=$BLAST:$PATH                                                        #
export PATH=$UTILS:$PATH                                                        #
export PATH=$DECODER:$PATH                                                      #
export PATH=$TRINOTATE:$PATH                                                    #
export PATH=$NOTATE_AD:$PATH                                                    #
export PATH=$PERL:$PATH                                                         #
export PATH=$AUTO:$PATH                                                         #
export PATH=$HMMER:$PATH
export PATH=$HMMSEARCH_BINARY:$PATH

```

